### PR TITLE
bpo-34042: Fix dict.copy() to maintain correct total refcount

### PR DESCRIPTION
**Pull request by 1st1**
https://bugs.python.org/issue34042

### CODE REVIEW

File: Lib/test/test_dict.py [hunk 1]
```diff
@@ -299,6 +299,24 @@ def test_copy_fuzz(self):
...
+                         'debug build required')
+    def test_copy_global_refcount(self):
+        # See issue #34042 for more details.
+        first_ref_count = second_ref_count = None
+        gettotalrefcount = sys.gettotalrefcount
+        dct = {'a': 1}
+
+        for i in range(10):
+            dct.copy()
+            first_ref_count = gettotalrefcount()
```
Comment (serhiy-storchaka) [inline Lib/test/test_dict.py]: Should it be in the loop?

File: Lib/test/test_dict.py [hunk 2]
```diff
@@ -299,6 +299,24 @@ def test_copy_fuzz(self):
                 self.assertNotEqual(d, d2)
                 self.assertEqual(len(d2), len(d) + 1)
 
+    @unittest.skipUnless(hasattr(sys, 'gettotalrefcount'),
+                         'debug build required')
+    def test_copy_global_refcount(self):
+        # See issue #34042 for more details.
+        first_ref_count = second_ref_count = None
```
Comment (serhiy-storchaka) [inline Lib/test/test_dict.py]: What is the reason of this initialization?

File: Lib/test/test_dict.py (hunk 1 above)
Comment (vstinner) [inline Lib/test/test_dict.py]: Why getting the total for each iteration of the loop?

**Comment by 1st1** [comment]
@vstinner @serhiy-storchaka The test is structured in a weird way because that's the only way how I can get it stable on my machine. And yet, here's a CI failure:

```
2018-07-05T20:27:40.8742568Z test test_dict failed -- Traceback (most recent call last):
2018-07-05T20:27:40.8766139Z   File "/opt/vsts/work/1/s/Lib/test/test_dict.py", line 318, in test_copy_global_refcount
2018-07-05T20:27:40.8783603Z     self.assertEqual(first_ref_count, second_ref_count)
2018-07-05T20:27:40.8798364Z AssertionError: 344832 != 344838
```

Any suggestions on how I can write a reliable test for `sys.gettotalrefcount()`?  Maybe commit this fix without a test?

**Comment by methane** [comment]
> Any suggestions on how I can write a reliable test for sys.gettotalrefcount()? Maybe commit this fix without a test?

Since this is debug feature, I'm OK for this instead of maintain cryptic, fragile test.

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
Please don't use fragile tests. I already have enough issues with existing fragile tests.

+1 to commit without tests.

**Comment by vstinner** [comment]
> What about using assertAlmostEqual with a delta much less than the number of loops?

Honestly, Python internals are too complex to get a reliable *total* reference count. I'm ok to remove the test.

regrtest usually runs a test 3 times to "warmup caches" before checking for reference leaks, and also use many low-level "cleanup" functions to help to get reliable reference counts.

**Comment by 1st1** [comment]
> Please don't use fragile tests. I already have enough issues with existing fragile tests.

I kind of like Serhiy's idea, but on the other hand Victor is the one who spends so much time herding our buildbots. I'll commit this PR without a test as @vstinner and @methane suggested.

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**